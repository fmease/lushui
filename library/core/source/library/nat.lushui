;; Natural numbers and related functions.
module

use topmost.type.Type
use topmost.(option.Option bool.Bool text.Text)

@public @intrinsic
data Nat: Type

@public
use add as +

@public @intrinsic
add: Nat -> Nat -> Nat

@public
use subtract as -

@public @intrinsic
subtract: Nat -> Nat -> Option Nat

@public
use panicking-subtract as -?

;;; @Temporary
@public @intrinsic
panicking-subtract: Nat -> Nat -> Nat

@public
use multiply as *

@public @intrinsic
multiply: Nat -> Nat -> Nat

@public
use divide as /

@public @intrinsic
divide: Nat -> Nat -> Option Nat

@public
use equal as ==

@public @intrinsic
equal: Nat -> Nat -> Bool

;;; equal (n m: Nat): Bool =
;;;     case duple Nat Nat n m of
;;;         duple Nat Nat \n n => true
;;;         \_ => false

@public
use less as <

@public @intrinsic
less: Nat -> Nat -> Bool

@public
use less-equal as =<

@public @intrinsic
less-equal: Nat -> Nat -> Bool

@public
use greater as >

@public @intrinsic
greater: Nat -> Nat -> Bool

@public
use greater-equal as >=

@public @intrinsic
greater-equal: Nat -> Nat -> Bool

@public @intrinsic
display: Nat -> Text

use topmost.(bool ordering)

@public
compare (n: Nat) (m: Nat): ordering.Ordering =
    case equal n m of
        bool.true => ordering.equal
        bool.false => case less n m of
            bool.true => ordering.less
            bool.false => ordering.greater

;;; @public @intrinsic
;;; to-int: Nat -> topmost.int.Int
