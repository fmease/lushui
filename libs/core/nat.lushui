use super.(option.Option bool.Bool text.Text)

@public @foreign
data Nat: Type

@public
use add as +

@public @foreign
add: Nat -> Nat -> Nat

@public
use subtract as -

@public @foreign
subtract: Nat -> Nat -> Option Nat

@public
use panicking-subtract as -?

;; @Temporary
@public @foreign
panicking-subtract: Nat -> Nat -> Nat

@public
use multiply as *

@public @foreign
multiply: Nat -> Nat -> Nat

@public
use divide as /

@public @foreign
divide: Nat -> Nat -> Option Nat

@public
use equal as ==

@public @foreign
equal: Nat -> Nat -> Bool

;; equal (n m: Nat): Bool =
;;     case duple Nat Nat n m of
;;         duple Nat Nat \n n => true
;;         \_ => false

@public
use less as <

@public @foreign
less: Nat -> Nat -> Bool

@public
use less-equal as =<

@public @foreign
less-equal: Nat -> Nat -> Bool

@public
use greater as >

@public @foreign
greater: Nat -> Nat -> Bool

@public
use greater-equal as >=

@public @foreign
greater-equal: Nat -> Nat -> Bool

@public @foreign
display: Nat -> Text

@public
use super.(bool ordering)

@public
compare (n m: Nat): ordering.Ordering =
    case equal n m of
        bool.true => ordering.equal
        bool.false => case less n m of
            bool.true => ordering.less
            bool.false => ordering.greater

;; @public @foreign
;; to-int: Nat -> Nat -> core.int.Int
