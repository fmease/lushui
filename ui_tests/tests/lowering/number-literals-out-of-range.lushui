;;; TEST fail check

use crates.core.nat.Nat

unleet: Nat = -1337

weird: Careless =
    accumulate @Nat64 1'000'000'000'000'000'000'000 @Int32 -9'999'999'999

@(recursion-limit 130'454'112'009'021)
module dummy of
