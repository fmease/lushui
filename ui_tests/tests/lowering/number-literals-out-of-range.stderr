error[E007]: number literal `-1337` does not fit type `Nat`
 --> ${DIRECTORY}/lowering/number-literals-out-of-range.lushui:5:15
  |
5 | unleet: Nat = -1337
  |               ^^^^^ 
  |
 note: values of this type must fit integer interval [0, infinity)

error[E007]: number literal `1000000000000000000000` does not fit type `Nat64`
 --> ${DIRECTORY}/lowering/number-literals-out-of-range.lushui:8:23
  |
8 |     accumulate @Nat64 1'000'000'000'000'000'000'000 @Int32 -9'999'999'999
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
  |
 note: values of this type must fit integer interval [0, 2^64-1]

error[E007]: number literal `-9999999999` does not fit type `Int32`
 --> ${DIRECTORY}/lowering/number-literals-out-of-range.lushui:8:60
  |
8 |     accumulate @Nat64 1'000'000'000'000'000'000'000 @Int32 -9'999'999'999
  |                                                            ^^^^^^^^^^^^^^ 
  |
 note: values of this type must fit integer interval [-2^31, 2^31-1]

error[E008]: attribute argument does not fit integer interval [0, 2^32-1]
  --> ${DIRECTORY}/lowering/number-literals-out-of-range.lushui:10:19
   |
10 | @(recursion-limit 130'454'112'009'021)
   |                   ^^^^^^^^^^^^^^^^^^^ 
   |

error: aborting due to 4 previous errors
 note: the errors E007 and E008 have a detailed explanation
 help: run `lushui explain <codes...>` to view a selection of them

